react/leanShipping.js

```diff
@@ -296,35 +302,40 @@ export function getCombinedOption(
   )
 }
 
-const getSlasAvailableToEveryItem = (slasByItem) =>
-  intersection(...slasByItem.map(slas => slas.map(sla => sla.id)))
+function getBestSlasBy(arrayOfSlas, property, shouldPrioritizeConsistentSlas) {
+  const slasAvailableToEveryItem = intersection(
+    ...arrayOfSlas.map(slas => slas.map(sla => sla.id))
+  )
+
+  const sortingCriterias = []
 
-function getMinSlaBy(items, property) {
-  const slasAvailableToEveryItem = getSlasAvailableToEveryItem(items)
+  // "Consistent SLAs" stands for SLAs that are available to every item
+  if (shouldPrioritizeConsistentSlas) {
+    sortingCriterias.push(sla => !slasAvailableToEveryItem.includes(sla.id))
+  }
+  sortingCriterias.push(sla => sla[property])
+  sortingCriterias.push(sla => sla.id)
 
-  return items
-  // Ordering by consistent SLA's makes it our tiebreaker
-    .map(item => orderBy(item, sla => slasAvailableToEveryItem.some(slaId => sla.id !== slaId)))
-    .map(slas => minBy(slas, slasGroup => slasGroup[property]))
+  return arrayOfSlas.map(slas => sortBy(slas, sortingCriterias)[0])
```

Comment:
since you are only getting the first item of the sort result, maybe you'd only need to run through the array once to get the best result, instead of sorting.

not sure if im missing something though, and if the arrays arent large then this shouldnt be much of a problem anyway

Reply:
Well this isn't trivial. I thought that I would only need to swap `sortBy(...)[0]` with `minBy(...)`, but `minBy` does not accept a comparison array.

I tested the implementation with both `minBy` (with a custom comparison callback) and with `reduce`, but neither of those two looked good/readable. See the `reduce` one below

```
slas.reduce((best, current) => {
  if (!best) {
    return current
  }

  for (const criteria of sortingCriterias) {
    if (criteria(best) !== criteria(current)) {
      return criteria(best) < criteria(current) ? best : current
    }
  }

  return current
})
```

Now I'd rather lose some performance and use the battle-tested `sortBy` :^/
What do you think?

Reply:
in this case I would also eschew the functional approach and do a `for..of` loop, running through the array once and storing the best result outside.

sortBy is ok I think, it just depends on how often this is called and how large is the array. but the comment was more of a suggestion anyway, it should be fine either way